src/cart/crt_iv.c

```diff
@@ -1141,6 +1141,8 @@ crt_ivf_rpc_issue(d_rank_t dest_node, crt_iv_key_t *iv_key,
 			IV_DBG(iv_key, "added to kip_entry=%p\n", entry);
 			D_MUTEX_UNLOCK(&entry->kip_lock);
 			D_MUTEX_UNLOCK(&ivns_internal->cii_lock);
+			if (rc == 0) /* drop the ref crt_hdlr_iv_fetch_aux() */
+				IVNS_DECREF(ivns_internal);
```

Comment:
i dont have a very strong opinion on either way it is done. i think current way is cleaner; am not sure why it would matter whether we store cb_info vs copy it for purposes of ref counting; copying simply would be less optimal, but fundamentally we are storing/saving cb_info->ivns_internal and in my view as such we should be taking ref count on it, which we do.

Reply:
On line 1138-1139, crt_ivf_pending_request_add is given cb_info, which includes a IV NS reference in cb_info->ifc_ivns_internal. It then puts cb_info to a list. Because crt_ivf_rpc_issue doesn't free cb_info afterwards, crt_ivf_pending_request_add is effectively taking over cb_info, instead of copying it. Hence, it simply doesn't need to take any extra IV NS reference.

In other words, the IV NS reference here is owned by cb_info, not crt_ivf_pending_request_add or crt_ivf_rpc_issue. The decref of ifc_ivns_internal (which "ivns_internal" aliases) shall always happen together with the D_FREE of the cb_info.

Reply:
Hmm, this actually makes sense, let me update the patch.